rfcs/remote_channel.md

```diff
@@ -0,0 +1,616 @@
+the parent have a handle to the child. Similarly, cross-origin
+navigations with the `cross-origin-opener-policy` header appropriately
+set create a new browsing context group which is totally isolated from
+the prior context. Typically in implementations each browsing context
+group is assigned a unique OS-level process.
+
+This creates some problems for testing; because testharness tests are
+running in a single browsing context, and test that involves a context
+that is isolated from the context containing the test will be unable
+to communicate test results back to the harness using web platform
+APIs.
+
+Given the importance of isolation in the modern web platform, it's
+important to improve support for tests involving multiple browsing
+context groups in web-platform-tests. Otherwise test authors will
+either not write tests for these scenarios, or write them using
+browser-specific techniques.
+
+### Prior Art
+
+* Gecko tests able to provide cross-context communication using the
+  parent (i.e. browser UI) process as an intermediary. The
+  [SpecialPowers.spawn](https://searchfox.org/mozilla-central/source/testing/specialpowers/content/SpecialPowersChild.jsm#1547-1584)
+  API available to gecko tests allows running a function in another
+  context, using structured clone to pass the arguments
+  ([example](https://searchfox.org/mozilla-central/source/dom/tests/browser/browser_data_document_crossOriginIsolated.js#17)).
+
+* The
+  [cross-origin-opener-policy](https://github.com/web-platform-tests/wpt/tree/7b0ebaccc62b566a1965396e5be7bb2bc06f841f/html/cross-origin-opener-policy)
+  tests currently in web-platform-tests use a bespoke
+  [dispatcher.py](https://github.com/web-platform-tests/wpt/blob/7b0ebaccc62b566a1965396e5be7bb2bc06f841f/html/cross-origin-opener-policy/resources/dispatcher.py)
+  to allow cross-origin communication. This uses a queue (implemented
+  as a list) stored in the wpt server stash object as a message
+  channel. Each context is given a UUID in a URL parameter named
+  `uuid`, and uses HTTP GET requests to check for any message in the
+  queue stored in the stash entry with that UUID. Other contexts that
+  know the UUID can POST a message that's appended to the queue.
+
+* WebDriver also has the ability to communicate between different
+  contexts, and [RFC
+  89](https://github.com/web-platform-tests/rfcs/pull/89) proposed
+  adding an `execute_script` method to testdriver to enable running a
+  script in a different context. However the design of WebDriver and
+  wptrunner means that only one context can communicate using
+  WebDriver at a time, and for testharness we must always return
+  control to the test window, since WebDriver is also used to
+  communicate test results back to the harness. The single threaded
+  nature of WebDriver, and requirement that all testdriver actions go
+  via the main test window, make it hard to build an ergonomic
+  cross-context messaging API.
+
+### Proposal
```

Comment:
I'd like to clarify (somewhere in the RFC) what are the dependencies of this RFC in terms of implementation.

IIUC this RFC can be implementable by pure JavaScript outside WPT infra + existing server-side stash, which is good because we can run the tests by running wpt server and bare browsers, without further depending on WPT test infra.
This RFC mentions integration with WebDriver BiDi, test_driver and testharness, but these are out of scope of this RFC itself.

Is my understanding correct?
Why the changes to `resources/testharness.js` and `tools/wptrunner/wptrunner` are needed in https://github.com/web-platform-tests/wpt/pull/29803 ?

(Other changes, in JS/HTML files, `tools/wptserve/wptserve/stash.py`, and `websockets/handlers/msg_channel_wsh.py` look like "implementable by pure JavaScript outside WPT infra + existing server-side stash")